pkg/storage/sqlcommon/sqlcommon.go

```diff
@@ -728,3 +765,33 @@ func AddFromUlid(sb sq.SelectBuilder, fromUlid string, sortDescending bool) sq.S
 	}
 	return sb.Where(sq.Gt{"ulid": fromUlid})
 }
+
+// HandleSQLError processes an SQL error and converts it into a more
+// specific error type based on the nature of the SQL error.
+func HandleSQLError(err error, args ...interface{}) error {
+	if errors.Is(err, sql.ErrNoRows) {
+		return storage.ErrNotFound
+	}
+
+	// handle mysql
+	if me, ok := err.(*mysql.MySQLError); ok && me.Number == 1062 {
+		if len(args) > 0 {
+			if tk, ok := args[0].(*openfgav1.TupleKey); ok {
+				return storage.InvalidWriteInputError(tk, openfgav1.TupleOperation_TUPLE_OPERATION_WRITE)
+			}
+		}
+		return storage.ErrCollision
+	}
+
+	// handle psql
+	if strings.Contains(err.Error(), "duplicate key value") {
+		if len(args) > 0 {
+			if tk, ok := args[0].(*openfgav1.TupleKey); ok {
+				return storage.InvalidWriteInputError(tk, openfgav1.TupleOperation_TUPLE_OPERATION_WRITE)
+			}
+		}
+		return storage.ErrCollision
+	}
+
```

Comment:
This was separated in #1615 so that we keep database-specific code in the individual storage plugins, is there a reason it needs to be combined together here again?

Reply:
The query execution has been moved outside of the specific DB driver into the SQLcommon directory. Thus these error handlers need to be generic, not just pushing the errors up as they were originally.

Reply:
There are numerous ways of solving this that do not rely on moving database-specific code into sqlcommon and reducing the pluggability of the individual stores.

Unilaterally resolving the comment and pushing the PR through without further discussion is a surprise to me, and I would very much like to see this feedback addressed in a follow-up PR to revert these changes and maintain the separation between sqlcommon and the database-specific code.